lib/posix-fdtab/fmap.h

```diff
@@ -0,0 +1,302 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+/* Lock-free data structure specialized in mapping integers to pointers */
+
+#ifndef __UK_FDTAB_FMAP_H__
+#define __UK_FDTAB_FMAP_H__ 1
+
+#include <string.h>
+
+#include <uk/arch/atomic.h>
+#include <uk/assert.h>
+#include <uk/bitops.h>
+#include <uk/bitmap.h>
+#include <uk/thread.h>
+
+/**
+ * Lock-free bitmap, with ones representing a free index.
+ */
+struct uk_bmap {
+	size_t size;
+	volatile unsigned long *bitmap;
+};
+
+#define UK_BMAP_SZ(s) (UK_BITS_TO_LONGS(s) * sizeof(unsigned long))
+
+
+/**
+ * Initialize the bitmap; must not be called concurrently with other functions.
+ */
+static inline void uk_bmap_init(const struct uk_bmap *bm)
+{
+	uk_bitmap_fill((void *)bm->bitmap, bm->size);
+}
+
+/**
+ * Mark `idx` as used, and return whether we were the ones to do so.
+ *
+ * @return
+ * 	0 if we marked `idx` as used, non-zero otherwise
+ */
+static inline int uk_bmap_reserve(const struct uk_bmap *bm, size_t idx)
```

Comment:
Leftover from a more generic design; can convert into int. Although in this particular case is makes no difference since the idx is not stored anywhere in memory.